crates/video/src/main.rs

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+mod stream;
+mod vhu_video;
+mod vhu_video_thread;
+mod video;
+mod video_backends;
+
+use std::path::Path;
+use std::sync::{Arc, RwLock};
+
+use clap::Parser;
+use log::{info, warn};
+use thiserror::Error as ThisError;
+use vhost::{vhost_user, vhost_user::Listener};
+use vhost_user_backend::VhostUserDaemon;
+use vhu_video::{BackendType, VuVideoBackend};
+use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
+
+pub(crate) type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug, ThisError)]
+pub(crate) enum Error {
+    #[error("Could not create backend: {0}")]
+    CouldNotCreateBackend(vhu_video::VuVideoError),
+    #[error("Could not create daemon: {0}")]
+    CouldNotCreateDaemon(vhost_user_backend::Error),
+    #[error("Failed creating listener: {0}")]
+    FailedCreatingListener(vhost_user::Error),
+}
+
+#[derive(Clone, Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct VideoArgs {
+    /// Unix socket to which a hypervisor connects to and sets up the control path with the device.
+    #[clap(short, long)]
+    socket_path: String,
+
+    /// Path to the video device file. Defaults to /dev/video0.
+    #[clap(short = 'd', long, default_value = "/dev/video0")]
+    v4l2_device: String,
+
+    /// Video backend to be used.
+    #[clap(short, long)]
+    #[clap(value_enum)]
+    backend: BackendType,
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) struct VuVideoConfig {
+    pub socket_path: String,
+    pub v4l2_device: String,
+    pub backend: BackendType,
+}
+
+impl TryFrom<VideoArgs> for VuVideoConfig {
+    type Error = Error;
+
+    fn try_from(args: VideoArgs) -> Result<Self> {
+        // Divide available bandwidth by the number of threads in order
+        // to avoid overwhelming the HW.
+        let socket_path: String = args.socket_path.trim().to_string();
+        let v4l2_device: String = args.v4l2_device.trim().to_string();
+        let backend: BackendType = args.backend;
+
+        Ok(VuVideoConfig {
+            socket_path,
+            v4l2_device,
+            backend,
+        })
+    }
+}
+
+impl VuVideoConfig {
+    /// Return the path of the unix domain socket which is listening to
+    /// requests from the guest.
+    pub fn get_socket_path(&self) -> String {
+        String::from(&self.socket_path)
```

Comment:
```suggestion
    pub fn get_socket_path(&self) -> &str {
        &self.socket_path
```

Reply:
Most crates use `String`, yes, I just copied for consistency. I will change it by `PathBuf` as in scsi.